hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/CoalesceFunction.java

```diff
@@ -0,0 +1,147 @@
+        return SqlOperandCountRanges.any();
+    }
+
+    @Override
+    protected boolean checkOperandTypes(HazelcastCallBinding callBinding, boolean throwOnFailure) {
+        HazelcastSqlValidator validator = callBinding.getValidator();
+        SqlValidatorScope scope = callBinding.getScope();
+
+        SqlCall call = callBinding.getCall();
+        List<SqlNode> operandList = call.getOperandList();
+
+        List<RelDataType> argTypes = new ArrayList<>(operandList.size());
+
+        boolean foundNotNull = false;
+        for (SqlNode node : operandList) {
+            argTypes.add(validator.deriveType(scope, node));
+            foundNotNull |= !SqlUtil.isNullLiteral(node, false);
+        }
+
+        if (!foundNotNull) {
+            if (throwOnFailure) {
+                throw callBinding.newError(RESOURCE.mustNotNullInElse());
+            }
+            return false;
+        }
+
+        RelDataType operandType = argTypes.get(0);
+        for (int i = 1; i < argTypes.size(); i++) {
+            operandType = HazelcastTypeUtils.withHigherPrecedence(operandType, argTypes.get(i));
+        }
+
+        QueryDataType caseHzReturnType = HazelcastTypeUtils.toHazelcastType(operandType.getSqlTypeName());
+
+        if (!allBranchTypesCanBeConvertedToReturnType(argTypes, operandList, operandType, caseHzReturnType)) {
+            if (throwOnFailure) {
+                throw QueryException.error(SqlErrorCode.GENERIC, "Cannot infer return type for COALESCE among " + argTypes);
+            } else {
+                return false;
+            }
+        }
+
+        for (int index = 0; index < operandList.size(); index++) {
+            validator.getTypeCoercion().coerceOperandType(scope, callBinding.getCall(), index, operandType);
+        }
+
+        return true;
+    }
+
+
+    private boolean allBranchTypesCanBeConvertedToReturnType(
+            List<RelDataType> argTypes,
+            List<SqlNode> allReturnNodes,
+            RelDataType caseReturnType,
+            QueryDataType caseHzReturnType) {
+        for (int i = 0, argTypesSize = argTypes.size(); i < argTypesSize; i++) {
+            RelDataType type = argTypes.get(i);
+            QueryDataType hzType = HazelcastTypeUtils.toHazelcastType(type.getSqlTypeName());
+
+            SqlNode sqlNode = allReturnNodes.get(i);
+
+            if (!(hzType.getTypeFamily() == QueryDataTypeFamily.NULL
+                    || type.equals(caseReturnType)
+                    || bothParametersAreNumeric(caseHzReturnType, hzType)
+                    || bothOperandsAreTemporalAndLowOperandCanBeConvertedToHighOperand(caseHzReturnType, hzType)
+                    || highOperandIsTemporalAndLowOperandIsLiteralOfVarcharType(caseHzReturnType, hzType, sqlNode))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean bothParametersAreNumeric(QueryDataType highHZType, QueryDataType lowHZType) {
```

Comment:
We already have exactly same code in `HazelcastTypeCoercion`, maybe it's worth share it?